### PR TITLE
Auth: Fix Last Seen being updated on every request

### PR DESCRIPTION
**Pull request by Jguer**
**What is this bug?**

Since `SyncLastSeenHook` runs after `FetchSyncedUserHook` 

when `if !shouldUpdateLastSeen(identity.LastSeenAt) {` is called the LastSeen of the identity is always empty and therefore it will always return `true` causing `s.userService.UpdateLastSeenAt` to be called.

**How do you fix it**

_First commit_:

We change the priority of the hook from 40 to 120, to make sure it runs after `FetchSyncedUserHook` (priority 110) that populates the identity struct before it's returned

_Second commit_:

A hook should not depend on another hook's behavior to function though.

`userService.UpdateLastSeenAt` should always be called and it should check if it needs to update the last seen at instead. 

Attempt to use a user retrieval with cache to avoid getting the same user multiple times. (cache time is 5s)

**Notes to reviewer**:

Second commit can be dropped and the bug will still be fixed

**What's missing**

- [ ] integration test to avoid regression
- [ ] user service method needs a unit test

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/authn/authnimpl/service.go [hunk 1]
```diff
@@ -155,7 +155,7 @@ func ProvideService(
 	s.RegisterPostAuthHook(userSyncService.SyncUserHook, 10)
 	s.RegisterPostAuthHook(userSyncService.EnableDisabledUserHook, 20)
 	s.RegisterPostAuthHook(orgUserSyncService.SyncOrgRolesHook, 30)
-	s.RegisterPostAuthHook(userSyncService.SyncLastSeenHook, 40)
+	s.RegisterPostAuthHook(userSyncService.SyncLastSeenHook, 120)
```
Comment (gamab) [inline pkg/services/authn/authnimpl/service.go:158]:
```suggestion
	// Hook needs to run after FetchSyncedUserHook (priority 110) to have a populated identity struct
	s.RegisterPostAuthHook(userSyncService.SyncLastSeenHook, 120)
```

**Review by gamab — APPROVED** [review]

File: pkg/services/authn/authnimpl/service.go (hunk 1 above)
Comment (Jguer) [reply pkg/services/authn/authnimpl/service.go:158]: with the second commit it actually doesn't need to run after `FetchSyncedUserHook`

**Review by alexanderzobnin — APPROVED** [review]

File: pkg/services/contexthandler/contexthandler.go [hunk 2]
```diff
@@ -242,7 +242,7 @@ func (h *ContextHandler) Middleware(next http.Handler) http.Handler {
 			// update last seen every 5min
 			if reqContext.ShouldUpdateLastSeenAt() {
 				reqContext.Logger.Debug("Updating last user_seen_at", "user_id", reqContext.UserID)
-				if err := h.userService.UpdateLastSeenAt(mContext.Req.Context(), &user.UpdateUserLastSeenAtCommand{UserID: reqContext.UserID}); err != nil {
+				if err := h.userService.UpdateLastSeenAt(mContext.Req.Context(), &user.UpdateUserLastSeenAtCommand{UserID: reqContext.UserID, OrgID: reqContext.OrgID}); err != nil {
```
Comment (gamab) [inline pkg/services/contexthandler/contexthandler.go:245]:
```suggestion
				err := h.userService.UpdateLastSeenAt(mContext.Req.Context(), &user.UpdateUserLastSeenAtCommand{UserID: reqContext.UserID, OrgID: reqContext.OrgID});
				if err != nil && !errors.Is(err, user.ErrLastSeenUpToDate) {
```

**Review by gamab — APPROVED** [review]
LGTM! Thanks again for the explanations, very interesting.

File: pkg/services/user/userimpl/user.go [hunk 3]
```diff
@@ -219,9 +219,26 @@ func (s *Service) ChangePassword(ctx context.Context, cmd *user.ChangeUserPasswo
...
+
+	if !shouldUpdateLastSeen(u.LastSeenAt) {
+		return user.ErrLastSeenUpToDate
+	}
+
 	return s.store.UpdateLastSeenAt(ctx, cmd)
 }
 
+func shouldUpdateLastSeen(t time.Time) bool {
+	return time.Since(t) > time.Minute*5
```
Comment (linoman) [inline pkg/services/user/userimpl/user.go:239]: We don't have a constant for these 5 minutes, don't we?

**Review by linoman — APPROVED** [review]
Looks great! I left a small non-blocking comment

**Review by leandro-deveikis — APPROVED** [review]
LGTM! :rocket:

**Pull request merged**